lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,242 @@
+            path accessMode
+        setMode600 path
+#endif
+
+-- | Create user public file at the given path, but only when one doesn't
+-- already exist.
+initializeUserPublic :: MonadMaybeLog m => FilePath -> m ()
+initializeUserPublic publicPath = do
+    exists <- liftIO $ doesFileExist publicPath
+#ifdef POSIX
+    if exists
+    then ensureModeIs600 publicPath
+    else do
+        createEmptyFile publicPath
+        setMode600 publicPath
+#else
+    unless exists $ createEmptyFile publicPath
+#endif
+  where
+    createEmptyFile :: (MonadIO m) => FilePath -> m ()
+    createEmptyFile = liftIO . flip writeFile mempty
+
+-- | Reads user public from the given file.
+-- If the file does not exist/is empty, returns empty user public
+peekUserPublic :: MonadMaybeLog m => FilePath -> m UserPublic
+peekUserPublic path = do
+    initializeUserPublic path
+    takeReadLock path $ do
+        content <- decodeFull' <$> BS.readFile path
+        pure $ either (const def) identity content & upPath .~ path
+
+-- | Read user public putting an exclusive lock on it. To unlock, use
+-- 'writeUserPublicRelease'.
+takeUserPublic :: MonadMaybeLog m => FilePath -> m UserPublic
+takeUserPublic path = do
+    initializeUserPublic path
+    liftIO $ do
+        lock <- lockFile (lockFilePath path) Exclusive
+        content <- decodeFull' <$> BS.readFile path
+        pure $ either (const def) identity content
+            & upPath .~ path
+            & upLock .~ Just lock
+
+-- | Writes user public.
+writeUserPublic :: (MonadIO m) => UserPublic -> m ()
+writeUserPublic up
+    | canWrite up = liftIO $ throwM $ KeyError Public AlreadyLocked
+    | otherwise   = liftIO $ withFileLock (lockFilePath $ up ^. upPath) Exclusive $ const $ writeRaw up
+
+-- | Writes user public and releases the lock. UserPublic can't be
+-- used after this function call anymore.
+writeUserPublicRelease :: (MonadIO m, MonadThrow m) => UserPublic -> m ()
+writeUserPublicRelease up
+    | not (canWrite up) = throwM $ KeyError Public NotWritable
+    | otherwise = liftIO $ do
+        writeRaw up
+        case (up ^. upLock) of
+            Nothing   -> throwM $ KeyError Public IncorrectLock
+            Just lock -> unlockFile lock
+
+-- | Helper for writing public to file
+writeRaw :: UserPublic -> IO ()
+writeRaw up = do
+    let path = up ^. upPath
+    -- On POSIX platforms, openBinaryTempFile guarantees that the file
+    -- will be created with mode 600.
+    -- If openBinaryTempFile throws, we want to propagate this exception,
+    -- hence no handler.
+    (tempPath, tempHandle) <-
+        openBinaryTempFile (takeDirectory path) (takeFileName path)
+
+    BS.hPut tempHandle (serialize' up) `finally` hClose tempHandle
+
+    renameFile tempPath path
+
+-- | Helper for taking shared lock on file
+takeReadLock :: MonadIO m => FilePath -> IO a -> m a
+takeReadLock path = liftIO . withFileLock (lockFilePath path) Shared . const
```

Comment:
Really minor quibble here, the name `takeReadLock` suggest that this function takes the lock. Maybe `withReadLock` would be a better name, like `withFile`